core/src/test/scala/com/mesosphere/usi/core/scalaapi/SchedulerAdapterTest.scala

```diff
@@ -0,0 +1,315 @@
+package com.mesosphere.usi.core.scalaapi
+
+import akka.NotUsed
+import akka.stream.scaladsl.{SinkQueueWithCancel, SourceQueueWithComplete}
+import akka.stream.scaladsl.{Flow, Sink, Source}
+import akka.stream.OverflowStrategy
+import com.mesosphere.usi.core.Scheduler
+import com.mesosphere.usi.core.models.resources.ScalarRequirement
+import com.mesosphere.usi.core.models._
+import com.mesosphere.utils.AkkaUnitTest
+import org.scalatest._
+
+import scala.concurrent.Promise
+import scala.util.{Failure, Success}
+
+class SchedulerAdapterTest extends AkkaUnitTest with Inside {
+
+  def createMockedScheduler: (
+      Flow[Scheduler.SpecInput, Scheduler.StateOutput, NotUsed],
+      SinkQueueWithCancel[Scheduler.SpecInput],
+      SourceQueueWithComplete[Scheduler.StateOutput]) = {
+
+    val (outerSinkQueue, outerSink) = Sink.queue[Scheduler.SpecInput]().preMaterialize()
+    val (outerSourceQueue, outerSource) =
+      Source.queue[Scheduler.StateOutput](16, OverflowStrategy.fail).preMaterialize()
+
+    (Flow.fromSinkAndSource(outerSink, outerSource), outerSinkQueue, outerSourceQueue)
+  }
+
+  class SchedulerFixture {
+    val podSpec = PodSpec(
+      PodId("podId"),
+      Goal.Running,
+      RunSpec(
+        resourceRequirements = List(ScalarRequirement.cpus(1), ScalarRequirement.memory(256)),
+        shellCommand = "sleep 3600")
+    )
+    val specEvent = PodSpecUpdated(podSpec.id, None)
+    val stateEvent = PodStatusUpdated(podSpec.id, None)
+
+    val (scheduler, specInputQueue, stateOutputQueue) = createMockedScheduler
+    val adapter = new SchedulerAdapter(scheduler)
+  }
+
+  "SchedulerAdapter" when {
+    "materialized as Source and Sink" should {
+      "Propagate a snapshot to the underlying scheduler" in new SchedulerFixture {
+        adapter.asSourceAndSink(SpecsSnapshot(List(podSpec), Nil))
+        specInputQueue.pull().futureValue.get._1.podSpecs.head shouldEqual podSpec
+      }
+
+      "If snapshot is not provided push an empty snapshot" in new SchedulerFixture {
+        adapter.asSourceAndSink()
+        specInputQueue.pull().futureValue.get._1 shouldEqual SpecsSnapshot.empty
+      }
+
+      "Complete the snapshot promise once it's available" in new SchedulerFixture {
+        val (snapshotF, _, _) = adapter.asSourceAndSink()
+        specInputQueue.pull().futureValue
+
+        val stateSnapshot = StateSnapshot(List(PodStatus(podSpec.id, Map.empty)), Nil, Nil, Nil)
+
+        stateOutputQueue.offer(stateSnapshot -> Source.maybe)
+
+        snapshotF.futureValue shouldEqual stateSnapshot
+      }
+
+      "Push the SpecUpdatedEvents downstream" in new SchedulerFixture {
+        val (_, _, sink) = adapter.asSourceAndSink()
+
+        Source.repeat(specEvent).runWith(sink)
+
+        specInputQueue.pull().futureValue.get._2.runWith(Sink.head).futureValue shouldEqual specEvent
```

Comment:
Ah, could you use the following syntax

```scala
"Push the SpecUpdatedEvents downstream" in {
  val f = new SchedulerFixture

  f.specInputQueue.pull().futureValue ...
}
```

it was not obvious to me that you are using a fixture here.

Reply:
let's deconstruct

```
Source.repeat(specEvent).runWith(sink)

val done = inside(specInputQueue.pull().futureValue) {
  case (_, events) => events.runWith(Sink.head)
}

done.futureValue shouldEqual specEvent
```

etc

Reply:
@jeschkies I personally find `f.method_call` pattern a bit verbose. Do you think we should agree to use/not use it? For example, in Marathon, we use `in new Fixture {` pattern in many tests.